src/Dialect/ONNX/Math/ElementwiseUnary.cpp

```diff
@@ -0,0 +1,778 @@
+}
+
+//===----------------------------------------------------------------------===//
+// ErfOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXErfOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  return inferShapeForUnaryElementwiseOps(this->getOperation());
+}
+
+//===----------------------------------------------------------------------===//
+// Exp
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXExpOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  return inferShapeForUnaryElementwiseOps(this->getOperation());
+}
+
+//===----------------------------------------------------------------------===//
+// Floor
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXFloorOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  return inferShapeForUnaryElementwiseOps(this->getOperation());
+}
+
+//===----------------------------------------------------------------------===//
+// HardSigmoid
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXHardSigmoidOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  return inferShapeForUnaryElementwiseOps(this->getOperation());
+}
+
+//===----------------------------------------------------------------------===//
+// HardmaxOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXHardmaxOp::verify() {
+  ONNXHardmaxOpAdaptor operandAdaptor(*this);
+  Value input = operandAdaptor.input();
+  if (!hasShapeAndRank(input))
+    return success(); // Won't be able to do any checking at this stage.
+
+  // axis attribute must be in the range [-r,r-1], where r = rank(input).
+  int64_t axisValue = axis();
+  int64_t inputRank = input.getType().cast<ShapedType>().getRank();
+  if (axisValue < -inputRank || axisValue >= inputRank)
+    return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
+        *this->getOperation(), "axis", axisValue,
+        onnx_mlir::Diagnostic::Range<int64_t>(-inputRank, inputRank - 1));
+
+  return success();
+}
+
+LogicalResult ONNXHardmaxOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  if (!hasShapeAndRank(input()))
+    return success();
+
+  auto inputType = input().getType().cast<ShapedType>();
+  int64_t inputRank = inputType.getRank();
+  int64_t axisValue = axis();
+
+  // axis attribute must be in the range [-r,r], where r = rank(input).
+  if (axisValue < -inputRank || axisValue > inputRank)
+    return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
+        *this->getOperation(), "axis", axisValue,
+        onnx_mlir::Diagnostic::Range<int64_t>(-inputRank, inputRank - 1));
+
+  return inferShapeForUnaryElementwiseOps(this->getOperation());
```

Comment:
I can see how it makes sense for Hardmax to share shape inference with conventional unary ops, but it's misleading that it is in a source file called "elementwise".

Should we rename the source fil to just "Unary.cpp" or break ops like Hardmax (and Scatter etc) into other files?

Reply:
Let me split them from the elementary